demo of full authorization flow

andorsk opened the pull request:
a sample demo with a full authorization flow. This is moved from https://github.com/TBD54566975/ssi-sdk/pull/123. In this, 3 actors are presented: 

1. A university
2. A student
3. An employer

The situation is simple: A student graduates from Example University, and Employer now wants to hire the student. This flow outlines how the VC flow works, between the holder, issuer, and verifier. Below is an example of the flow. 

<img width="1035" alt="image" src="https://user-images.githubusercontent.com/8604639/179478517-ff0cc2ca-b322-4c4a-ba88-cb240dc8d082.png">

There is a basic authorization check later, which checks if the employer is OK with the student's VC claims. This is intended to be an explanatory demo, not a production use case.

andorsk commented:
@decentralgabe can you take a look at this and let me know if at least the flow is correct and there's nothing wrong conceptually with the demonstration? In particular, the exchange layer is one that I think needs to be really clear, as that seems the layer that seems the most room for error.

decentralgabe commented on example/demo1/demo.go:
```diff
@@ -0,0 +1,700 @@
+//  1. DID for the Holder
+//  2. DID for the issuer
+//  3. VC for the Holder from the verifier
+//  4. PresentationRequest for the Verifier
```
should this be from the Verifier?

decentralgabe commented on example/demo1/demo.go:
```diff
@@ -0,0 +1,700 @@
+)
+
+var (
+	UNSUPPORTED_DID_ERROR = errors.New("Unsupported Method for DID")
```
nit: `UnsupportedDIDError`

decentralgabe commented on example/demo1/demo.go:
```diff
@@ -0,0 +1,700 @@
+func resolveDID(didStr string) (*did.DIDDocument, error) {
+	split := strings.Split(string(didStr), ":")
+	if len(split) < 2 {
+		return nil, errors.New("Invalid DID. Does not split correctly")
```
nit: no caps on errors `errors.New("invalid DID...`

decentralgabe commented on example/demo1/demo.go:
```diff
@@ -0,0 +1,700 @@
+	}
+	var method = split[1]
+	switch method {
+	case "key":
```
can likely use the known prefixes from the lib here

decentralgabe reviewed (commented):
will review in a bit more detail later this week. high level looks good. perhaps worth breaking out into files per party (e.g. holder, verifier, issuer) to make things a bit simpler to follow?

andorsk commented:
@decentralgabe yea...we can break it up into multiple files. Good idea. Thanks for the comments.

andorsk commented:
I need to eventually bring this back to life. Sorry, been caught up with other things.

andorsk commented:
Based on https://github.com/TBD54566975/ssi-sdk/pull/164 will move over to use_cases.

andorsk commented:
@decentralgabe I think this example will work. I've incorporated the feedback provided. There's thing that I would like to update and fix, in later PR's, but for now I'd like to just get this in, and improve the example later. 

Can you take a look when you have a chance, and we can discuss what changes would be needed for a merge? I'd like to limit future work on it, but I don't want to scrap this work either as it accomplishes I think, it's primary goal. 

<img width="1051" alt="image" src="https://user-images.githubusercontent.com/8604639/184726422-7c5b2a77-1d79-4261-97c9-d88a4756af4b.png">

andorsk commented:
@decentralgabe please check out the comment above re: my position on this PR.

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/issuer.go:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	err := knownCred.IsValid()
+	if err != nil {
```
nit one liner

`if err := knownCred.IsValid(); err != nil { `

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/issuer.go:
```diff
@@ -0,0 +1,83 @@
+	"github.com/sirupsen/logrus"
+)
+
+func BuildExampleUniversityVC(universityID string, recipient string) (*credential.VerifiableCredential, error) {
```
this doesn't have to be changed, but in case you didn't know if multiple parameters have the same type you can do this

`BuildExampleUniversityVC(universityID, recipient string)`

andorsk replied:
yep. I'm aware. Good catch yough.

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+	"github.com/sirupsen/logrus"
+)
+
+var cw = CustomStepWriter{}
```
generally don't like mutable state like this...is there another way we can initialize and pass it as the flow goes on?

andorsk replied:
resolved.

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+// but the resolution would need to be hooked with the sdk.
+//  TODO (andor): Should exist a universal resolution method somewhere
+// in the actual SDK
+func resolveDID(didStr string) (*did.DIDDocument, error) {
```
this would be a good addition to the `did` package

andorsk replied:
ahhhh! OK. I will add a PR with this in the ssi-sdk!

andorsk replied:
done.

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+		wallet: NewSimpleWallet(),
+		Name:   name,
+	}
+	err := e.wallet.Init(keyType)
```
one liner

andorsk replied:
Thanks!

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+}
+
+func InitSampleDID() (error, did.DID) {
+	return initSampleBody()
```
this is a little confusing, what is sample body doing?

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+	step int
+}
+
+func (i *CustomStepWriter) Write(s string) {
```
nit: usually make the receiver name somewhat-meaningful like (csw *CustomStepWriter)

andorsk replied:
sure thing.

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+func validateVC(vc credential.VerifiableCredential) error {
+
+	issuer := "https://example.edu/issuers/565049"
+	var AssertionMethod cryptosuite.ProofPurpose = "assertionMethod"
+	var vc2 credential.VerifiableCredential
```
if you're assigning a value, don't use var favor `:=`

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+	if err != nil {
+		return err
+	}
+	var OKP = cryptosuite.KTY("OKP")
```
here and below too

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+	issuer := "https://example.edu/issuers/565049"
+	var AssertionMethod cryptosuite.ProofPurpose = "assertionMethod"
+	var vc2 credential.VerifiableCredential
+	err := util.Copy(&vc, &vc2)
```
one line

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+	}
+	suite := cryptosuite.GetJSONWebSignature2020Suite()
+	err = suite.Sign(signer, &vc2)
+	if err != nil {
```
one line

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+	if err != nil {
+		return err
+	}
+	err = suite.Verify(verifier, &vc2)
```
one line

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+// and for the source code with the sdk,
+// https://github.com/TBD54566975/ssi-sdk/blob/main/credential/exchange/request.go
+// is appropriate to start off with.
+func makePresentationRequest(jwk cryptosuite.JSONWebKey2020, presentationData exchange.PresentationDefinition, targetId string) (pr []byte, signer *cryptosuite.JSONWebKeySigner, err error) {
```
nit: targetID

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+	return requestJWTBytes, signer, err
+}
+
+func MakePresentationRequest(jwk cryptosuite.JSONWebKey2020, presentationData exchange.PresentationDefinition, targetId string) (pr []byte, signer *cryptosuite.JSONWebKeySigner, err error) {
```
targetID

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+	return buildPresentationSubmission(presentationRequest, signer, verifier, vc)
+}
+
+// https://github.com/TBD54566975/ssi-sdk/blob/d279ca2779361091a70b8aa3c685a388067409a9/credential/exchange/submission.go#L126
```
maybe we should use relative references? meaning `/credential/exchange/submission.go`

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+		return nil, err
+	}
+	var pd exchange.PresentationDefinition
+	err = json.Unmarshal(dat, &pd)
```
one line

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+}
+
+func MakePresentationData(id string, inputID string) (exchange.PresentationDefinition, error) {
+	return makePresentationData(id, inputID)
```
can this just be one call?

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/util.go:
```diff
@@ -0,0 +1,315 @@
+
+func handleError(err error) {
+	if err != nil {
+		cw.WriteError(err.Error())
```
this assumes a CW is initialized, which is risky. perhaps this should take a cw or be a reciever on a CW

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/verifier.go:
```diff
@@ -0,0 +1,54 @@
+// 2. That the VC was issued by a trusted entity
+func validateAccess(verifier cryptosuite.JSONWebKeyVerifier, data []byte) error {
+
+	var authorizationError = errors.New("insufficient claims provided")
```
no var

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/verifier.go:
```diff
@@ -0,0 +1,54 @@
+		if dat, err := json.Marshal(untypedCredential); err == nil {
+
+			if err := json.Unmarshal(dat, &vc); err != nil {
+				logrus.Error(err)
```
is the intention to continue processing if this fails?

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/wallet.go:
```diff
@@ -0,0 +1,126 @@
+// private keys
+// and vCs
+type SimpleWallet struct {
+	vCs  map[string]*credential.VerifiableCredential
```
vcs

decentralgabe replied:
is there a reason to have a map to a pointer?

andorsk replied:
great question. No particular reason.

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/wallet.go:
```diff
@@ -0,0 +1,126 @@
+	mux  *sync.Mutex
+}
+
+func NewSimpleWallet() *SimpleWallet {
```
I like this wallet, perhaps move it up so it benefits all use cases?

andorsk replied:
👀 yea! sounds good

andorsk replied:
done.

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/wallet.go:
```diff
@@ -0,0 +1,126 @@
+// adds it to the registry
+func (s *SimpleWallet) Init(keyType string) error {
+
+	s.mux.Lock() // TODO: remove the muxes?
```
what's the value here?

andorsk replied:
in the muxing? preventing concurrent override of k:v to prevent duplicate creds. For example: if I concurrently added VC with 1, they would override eachother in the hashmap.

decentralgabe commented on example/use_cases/employer_university_flow/employer_university_flow/wallet.go:
```diff
@@ -0,0 +1,126 @@
+	var didStr string
+	var err error
+
+	if keyType == "peer" {
```
there should be a well known const string for "peer" in that directory. if not, please create it

andorsk replied:
yea..good call

decentralgabe commented on example/use_cases/employer_university_flow/main.go:
```diff
@@ -0,0 +1,183 @@
+var debug = os.Getenv("DEBUG")
+
+func init() {
+	if debug == "1" {
```
can define modes as consts? perhaps even a type
`type Mode string`

andorsk replied:
got it.

decentralgabe reviewed (changes requested):
mostly minor changes, some flow control questions

decentralgabe commented:
noticed a number of package names with underscores. I know there are some already with this pattern (which should be corrected) - but the go style guide advises avoiding it https://go.dev/blog/package-names

andorsk commented:
> noticed a number of package names with underscores. I know there are some already with this pattern (which should be corrected) - but the go style guide advises avoiding it https://go.dev/blog/package-names

Good call out. Refactoring.

andorsk commented:
Addressed comments:

<img width="1218" alt="image" src="https://user-images.githubusercontent.com/8604639/187525296-8d662adb-7bf8-4dea-b77b-615168d5727a.png">

decentralgabe commented on did/util.go:
```diff
@@ -14,6 +15,10 @@ import (
 )
 
+var (
+	UnsupportedDIDErorr = errors.New("unsupported Method for DID")
```
typo

decentralgabe commented on did/util.go:
```diff
@@ -144,3 +149,29 @@ func keyTypeToMultiCodec(kt crypto.KeyType) (multicodec.Code, error) {
+	if len(split) < 2 {
+		return nil, errors.New("invalid DID. Does not split correctly")
+	}
+	var method = split[1]
```
`method := split[1]`

andorsk replied:
cool.

decentralgabe commented on example/use_cases/employer_university_flow/main.go:
```diff
@@ -0,0 +1,177 @@
+)
+
+// Set to debug mode here
+var debug = os.Getenv("DEBUG")
```
I'm wondering if there's a place we can keep all env vars for the service. Perhaps the `config` file is the right place?

I think this is fine for now...but may become unwieldy as a pattern

andorsk replied:
yep. I'm going to leave it for now, but we can adjust it later if it makes sense with a refactor.

decentralgabe commented on example/use_cases/employer_university_flow/main.go:
```diff
@@ -0,0 +1,177 @@
+	presentationData, err := emp.MakePresentationData("test-id", "id-1")
+	util.HandleExampleError(err, "failed to create pd")
+	if dat, err := json.Marshal(presentationData); err == nil {
+		logrus.Debugf("Presentation Data:\n%v", string(dat))
```
if there's an error here, the program will continue?

andorsk replied:
yea...the error would be contained to marshaling data. 1) It shouldn't ever really happen ( a failure to marshal ) but I see it as a design choice not a bug if it decides to not exit. There could be cases, where json marshaling fails for some reason. 

I'll change it to exit but wanted to note the above.

decentralgabe commented on example/use_cases/employer_university_flow/main.go:
```diff
@@ -0,0 +1,177 @@
+	util.HandleExampleError(err, "failed to verify jwt")
+
+	if dat, err := json.Marshal(vp); err == nil {
+		logrus.Debugf("Submission:\n%v", string(dat))
```
curious why not `util.HandleExampleError` here?

andorsk replied:
see above comments.

decentralgabe commented on example/use_cases/employer_university_flow/pkg/issuer.go:
```diff
@@ -0,0 +1,76 @@
+	"github.com/sirupsen/logrus"
+)
+
+func BuildExampleUniversityVC(universityID string, recipient string) (*credential.VerifiableCredential, error) {
```
Can we remove this and make the one below public?

andorsk replied:
done.

decentralgabe commented on example/use_cases/employer_university_flow/pkg/issuer.go:
```diff
@@ -0,0 +1,76 @@
+		CredentialSubject: knownSubject,
+	}
+
+	err := knownCred.IsValid()
+	if err != nil {
+		return nil, err
+	}
```
```suggestion
	if err := knownCred.IsValid(); err != nil {
		return nil, err
	}
```

decentralgabe commented on example/use_cases/employer_university_flow/pkg/issuer.go:
```diff
@@ -0,0 +1,76 @@
+		return nil, err
+	}
+
+	if dat, err := json.Marshal(knownCred); err == nil {
```
what happens if this is an error? does the "VC issued" line still show?

andorsk replied:
I was thinking we just ignore errors that are logging/marshaling json errors, but we can handle them. I don't think they should happen often if at all.

decentralgabe commented on example/use_cases/employer_university_flow/pkg/util.go:
```diff
@@ -0,0 +1,240 @@
+func validateVC(vc credential.VerifiableCredential) error {
+
+	issuer := "https://example.edu/issuers/565049"
+	var AssertionMethod cryptosuite.ProofPurpose = "assertionMethod"
```
```suggestion
	cryptosuite.ProofPurpose := "assertionMethod"
```

decentralgabe commented on example/use_cases/employer_university_flow/pkg/util.go:
```diff
@@ -0,0 +1,240 @@
+	issuer := "https://example.edu/issuers/565049"
+	var AssertionMethod cryptosuite.ProofPurpose = "assertionMethod"
+	var vc2 credential.VerifiableCredential
+	err := util.Copy(&vc, &vc2)
+	if err != nil {
+		return err
+	}
+	var OKP = cryptosuite.KTY("OKP")
+	var Ed25519 = cryptosuite.CRV("Ed25519")
```
```suggestion
	if err := util.Copy(&vc, &vc2): err != nil {
		return err
	}
	okp := cryptosuite.KTY("OKP")
	ed25519 := cryptosuite.CRV("Ed25519")
```

decentralgabe commented on example/use_cases/employer_university_flow/pkg/util.go:
```diff
@@ -0,0 +1,240 @@
+	if err != nil {
+		return err
+	}
+	err = suite.Verify(verifier, &vc2)
+	if err != nil {
```
```suggestion
	if err = suite.Verify(verifier, &vc2); err != nil {
```

decentralgabe commented on example/use_cases/employer_university_flow/pkg/util.go:
```diff
@@ -0,0 +1,240 @@
+		return err
+	}
+
+	suite := cryptosuite.GetJSONWebSignature2020Suite()
+	err = suite.Sign(signer, &vc2)
+	if err != nil {
+		return err
+	}
```
```suggestion
	suite := cryptosuite.GetJSONWebSignature2020Suite()
	if err = suite.Sign(signer, &vc2); err != nil {
		return err
	}
```

decentralgabe commented on example/use_cases/employer_university_flow/pkg/util.go:
```diff
@@ -0,0 +1,240 @@
+	return requestJWTBytes, signer, err
+}
+
+func MakePresentationRequest(jwk cryptosuite.JSONWebKey2020, presentationData exchange.PresentationDefinition, targetId string) (pr []byte, signer *cryptosuite.JSONWebKeySigner, err error) {
```
What is the point of this?

andorsk replied:
I'm assuming the question is why private and public mirror functions. So the idea was originally keeping everything private. Then I realized that I wanted to make some public, but I didn't want to refactor the private function calls. I've also seen this done sometimes, to keep separation of concerns. Private for functionality, then a wrapper to expose it. If you need to hide the function again, it's trivial. 

That being said: The style of private functions with an external wrapper is probably not useful for this demo, so I've adjusted things.

decentralgabe commented on example/use_cases/employer_university_flow/pkg/util.go:
```diff
@@ -0,0 +1,240 @@
+	return normalizedClaims
+}
+
+func BuildPresentationSubmission(presentationRequest []byte, signer cryptosuite.Signer, verifier cryptosuite.JSONWebKeyVerifier, vc credential.VerifiableCredential) ([]byte, error) {
```
same here, can this be removed?

decentralgabe commented on example/use_cases/employer_university_flow/pkg/verifier.go:
```diff
@@ -0,0 +1,54 @@
+// 2. That the VC was issued by a trusted entity
+func validateAccess(verifier cryptosuite.JSONWebKeyVerifier, data []byte) error {
+
+	var authorizationError = errors.New("insufficient claims provided")
```
either raise to be outside the method or `authorizationError := errors.New(....`

decentralgabe commented on example/use_cases/employer_university_flow/pkg/verifier.go:
```diff
@@ -0,0 +1,54 @@
+				logrus.Error(err)
+			}
+
+			if issuer, ok := vc.CredentialSubject["id"]; ok {
+				if TrustedEntities.isTrusted(issuer.(string)) {
+					authorizationError = nil
+				}
+			}
```
```suggestion
			issuer, ok := vc.CredentialSubject["id"]
			if  ok && TrustedEntities.isTrusted(issuer.(string)) {
				authorizationError = nil
			}
```

decentralgabe commented on example/use_cases/employer_university_flow/pkg/verifier.go:
```diff
@@ -0,0 +1,54 @@
+	return authorizationError
+}
+
+func ValidateAccess(verifier cryptosuite.JSONWebKeyVerifier, data []byte) error {
```
no need for this

decentralgabe commented on example/wallet.go:
```diff
@@ -0,0 +1,131 @@
+// private keys
+// and vCs
+type SimpleWallet struct {
+	vCs  map[string]credential.VerifiableCredential
```
```suggestion
	vcs  map[string]credential.VerifiableCredential
```

decentralgabe commented on example/wallet.go:
```diff
@@ -0,0 +1,131 @@
+// This holds the assigned dids
+// private keys
+// and vCs
+type SimpleWallet struct {
```
this is great!

decentralgabe commented on example/wallet.go:
```diff
@@ -0,0 +1,131 @@
+	}
+
+	CustomWriter.WriteNote(fmt.Sprintf("DID for holder is: %s", didStr))
+	s.AddPrivateKey("main", privKey)
```
nit: what do you think about indexing the keys with integers? this could be simpler...but i also see some benefit in having named keys...

andorsk replied:
I think we could definitely expand the wallet or do some other types of indexing. Another project I'm working on does this. 

To me though, there's another scope of work around making good wallets, and the scope of this PR is to make just a wallet that is simple in functionality and gives the general "idea".  As long as it works, and can educate, at least IMO, it's sufficient for a PR, with a possible follow up on an issue about doing an adjustment. 

But ultimately your call. If you want, we can explore different wallet styles in the PR, but it might result in some extra time required for acceptance.

decentralgabe replied:
agree it's for a follow up

decentralgabe reviewed (changes requested):
closer!

andorsk commented on example/use_cases/employer_university_flow/main.go:
```diff
@@ -0,0 +1,177 @@
+	}
+
+	jwk, err := cryptosuite.GenerateJSONWebKey2020(cryptosuite.OKP, cryptosuite.Ed25519)
+	if err != nil {
```
ah. this is because I don't handle the errors here. I handle errors in a single place. This just propagates the error up to the parent function. Basically, propagate the errors until the end, and then determine handling methods. This is from my understanding, best practice: 

https://earthly.dev/blog/golang-errors/

andorsk commented:
@decentralgabe thanks for the review! Hopefully addressed the comments/concerns. Made a few notes about some thoughts. Lmk if there's anything else. Looking forward to getting this in.

decentralgabe commented on did/util.go:
```diff
@@ -144,3 +149,29 @@ func keyTypeToMultiCodec(kt crypto.KeyType) (multicodec.Code, error) {
+		did, _, _, err := DIDPeer(didStr).Resolve()
+		return did, err
+	default:
+		return nil, fmt.Errorf("%v. Got %v method", UnsupportedDIDErorr, method)
```
```suggestion
		return nil, fmt.Errorf("%v. Got %v method", UnsupportedDIDError, method)
```

decentralgabe reviewed (approved):
thanks for seeing this through!
